File: src/lib/avatars.ts
import alpaca from "../assets/avatars/Alpaca.jpg";
import arcticFox from "../assets/avatars/Arcticfox.jpg";
import bear from "../assets/avatars/Bear.jpg";
import cheetah from "../assets/avatars/Cheetah.jpg";
import cheetah2 from "../assets/avatars/Cheetah2.jpg";
import chimpanzee from "../assets/avatars/Chimpanzee.jpg";
import dolphin from "../assets/avatars/Dolphin.jpg";
import fennec from "../assets/avatars/Fennec.jpg";
import gazelle from "../assets/avatars/Gazelle.jpg";
import gazelle2 from "../assets/avatars/Gazelle2.jpg";
import giraffe from "../assets/avatars/Giraffe.jpg";
import koala from "../assets/avatars/Koala.jpg";
import llama from "../assets/avatars/Llama.jpg";
import lynx from "../assets/avatars/Lynx.jpg";
import panda from "../assets/avatars/Panda.jpg";
import raccoon from "../assets/avatars/Raccoon.jpg";
import redPanda from "../assets/avatars/Redpanda.jpg";
import redPanda2 from "../assets/avatars/Redpanda2.jpg";
import retriever from "../assets/avatars/Retriever.jpg";
import shiba from "../assets/avatars/Shiba.jpg";
import snowLeopard from "../assets/avatars/Snowleopard.jpg";
import snowLeopard2 from "../assets/avatars/Snowleopard2.jpg";
import tiger from "../assets/avatars/Tiger.jpg";
import wolf from "../assets/avatars/Wolf.jpg";

export type Avatar = {
  id: string;
  name: string;
  src: string;
};

export const AVATARS: Avatar[] = [
  { id: "alpaca", name: "Alpaca", src: alpaca },
  { id: "arctic-fox", name: "Arctic fox", src: arcticFox },
  { id: "bear", name: "Bear", src: bear },
  { id: "cheetah", name: "Cheetah", src: cheetah },
  { id: "cheetah-2", name: "Cheetah", src: cheetah2 },
  { id: "chimpanzee", name: "Chimpanzee", src: chimpanzee },
  { id: "dolphin", name: "Dolphin", src: dolphin },
  { id: "fennec", name: "Fennec", src: fennec },
  { id: "gazelle", name: "Gazelle", src: gazelle },
  { id: "gazelle-2", name: "Gazelle", src: gazelle2 },
  { id: "giraffe", name: "Giraffe", src: giraffe },
  { id: "koala", name: "Koala", src: koala },
  { id: "llama", name: "Llama", src: llama },
  { id: "lynx", name: "Lynx", src: lynx },
  { id: "panda", name: "Panda", src: panda },
  { id: "raccoon", name: "Raccoon", src: raccoon },
  { id: "red-panda", name: "Red panda", src: redPanda },
  { id: "red-panda-2", name: "Red panda", src: redPanda2 },
  { id: "retriever", name: "Retriever", src: retriever },
  { id: "shiba", name: "Shiba", src: shiba },
  { id: "snow-leopard", name: "Snow leopard", src: snowLeopard },
  { id: "snow-leopard-2", name: "Snow leopard", src: snowLeopard2 },
  { id: "tiger", name: "Tiger", src: tiger },
  { id: "wolf", name: "Wolf", src: wolf },
];

const byId = new Map(AVATARS.map((avatar) => [avatar.id, avatar]));

export function avatarById(id: string | undefined): Avatar | undefined {
  if (!id) return undefined;
  return byId.get(id);
}

File: src-tauri/src/chat/avatars.rs
//! Conversation faces. Ids match `src/lib/avatars.ts`.

use std::collections::HashSet;

pub struct Avatar {
    pub id: &'static str,
    pub name: &'static str,
}

/// Spoken name is the animal, not the filename. Numbered variants share a name.
pub const AVATARS: &[Avatar] = &[
    Avatar {
        id: "alpaca",
        name: "Alpaca",
    },
    Avatar {
        id: "arctic-fox",
        name: "Arctic fox",
    },
    Avatar {
        id: "bear",
        name: "Bear",
    },
    Avatar {
        id: "cheetah",
        name: "Cheetah",
    },
    Avatar {
        id: "cheetah-2",
        name: "Cheetah",
    },
    Avatar {
        id: "chimpanzee",
        name: "Chimpanzee",
    },
    Avatar {
        id: "dolphin",
        name: "Dolphin",
    },
    Avatar {
        id: "fennec",
        name: "Fennec",
    },
    Avatar {
        id: "gazelle",
        name: "Gazelle",
    },
    Avatar {
        id: "gazelle-2",
        name: "Gazelle",
    },
    Avatar {
        id: "giraffe",
        name: "Giraffe",
    },
    Avatar {
        id: "koala",
        name: "Koala",
    },
    Avatar {
        id: "llama",
        name: "Llama",
    },
    Avatar {
        id: "lynx",
        name: "Lynx",
    },
    Avatar {
        id: "panda",
        name: "Panda",
    },
    Avatar {
        id: "raccoon",
        name: "Raccoon",
    },
    Avatar {
        id: "red-panda",
        name: "Red panda",
    },
    Avatar {
        id: "red-panda-2",
        name: "Red panda",
    },
    Avatar {
        id: "retriever",
        name: "Retriever",
    },
    Avatar {
        id: "shiba",
        name: "Shiba",
    },
    Avatar {
        id: "snow-leopard",
        name: "Snow leopard",
    },
    Avatar {
        id: "snow-leopard-2",
        name: "Snow leopard",
    },
    Avatar {
        id: "tiger",
        name: "Tiger",
    },
    Avatar {
        id: "wolf",
        name: "Wolf",
    },
];

pub fn name_for(id: &str) -> Option<&'static str> {
    AVATARS
        .iter()
        .find(|avatar| avatar.id == id)
        .map(|a| a.name)
}

pub fn pick_id(thread_id: &str, used: &HashSet<String>) -> &'static str {
    let unused: Vec<&Avatar> = AVATARS
        .iter()
        .filter(|avatar| !used.contains(avatar.id))
        .collect();
    if unused.is_empty() {
        return AVATARS[hash_thread_id(thread_id) % AVATARS.len()].id;
    }
    unused[hash_thread_id(thread_id) % unused.len()].id
}

fn hash_thread_id(thread_id: &str) -> usize {
    let mut hash: u32 = 0;
    for byte in thread_id.bytes() {
        hash = hash.wrapping_mul(31).wrapping_add(u32::from(byte));
    }
    hash as usize
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn names_match_catalog_ids() {
        assert_eq!(name_for("cheetah"), Some("Cheetah"));
        assert_eq!(name_for("cheetah-2"), Some("Cheetah"));
        assert_eq!(name_for("arctic-fox"), Some("Arctic fox"));
        assert_eq!(name_for("nope"), None);
    }

    #[test]
    fn pick_skips_faces_already_in_use() {
        let used: HashSet<String> = AVATARS
            .iter()
            .skip(1)
            .map(|avatar| avatar.id.to_string())
            .collect();
        assert_eq!(pick_id("t_anything", &used), AVATARS[0].id);
    }

    #[test]
    fn pick_is_stable_when_every_face_is_taken() {
        let used: HashSet<String> = AVATARS.iter().map(|avatar| avatar.id.to_string()).collect();
        assert_eq!(pick_id("same-thread", &used), pick_id("same-thread", &used));
        assert!(name_for(pick_id("same-thread", &used)).is_some());
    }
}
